File: API/promptSend.php
<?php


function getRequestInfo() {
       return json_decode(file_get_contents('php://input'), false);
}


// Function to call OpenAI API
function callOpenAI($apiKey, $model, $messages) {
   $url = 'https://api.openai.com/v1/chat/completions';


   // Set up the headers
   $headers = [
       'Content-Type: application/json',
       'Authorization: Bearer ' . $apiKey,
   ];


   // Prepare the request data
   $data = [
       'model' => $model,
       'messages' => $messages,
   ];


   // Initialize cURL
   $ch = curl_init($url);
   curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
   curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
   curl_setopt($ch, CURLOPT_POST, true);
   curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));


   // Execute the request
   $response = curl_exec($ch);


   // Check for errors
   if (curl_errno($ch)) {
       echo 'Error: ' . curl_error($ch);
       return null;
   }


   // Close cURL
   curl_close($ch);

   // Remove all occurrences of ''' characters
   $cleaned_response = str_replace("```", "", $response);
   $cleaned_response = str_replace("json", "", $cleaned_response);


   // Decode the JSON response
   return json_decode($cleaned_response, true);
}


function sendResultInfoAsJson( $obj ) {
      header('Content-type: application/json');
      echo $obj;
}


$apiKey = "PLACEHOLDER";


$inData = getRequestInfo();


$input = $inData->pdfText;


// Define the messages
$messages = [
   ["role" => "system", "content" => "You are a helpful tutor. Generate flashcards based on the information provided."],
   ["role" => "user", "content" => "CREATE THE MAXIMUM AMOUNT OF FLASHCARDS YOU CAN in ONLY RAW JSON (without additional text and do NOT include ```json) with question, answer, (question and answer MUST be a maximum of 30 words long!!!) and hint values as objects in an array based on this information: ". $input],
];


// Call the OpenAI API
$completion = callOpenAI($apiKey, 'gpt-3.5-turbo', $messages);


if ($completion && isset($completion['choices'][0]['message']['content']))
{
   sendResultInfoAsJson($completion['choices'][0]['message']['content']);
} else
{
   $http_status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
   //echo $http_status;
   echo "API Error: " . $completion['error']['message'] . "\n";
   //echo "Error in response from OpenAI API.\n";
}
?>

File: API/promptSend.py
from openai import OpenAI
import json

client = OpenAI(
    # This is the default and can be omitted
    api_key="",
)
# Function to generate flashcards with questions, answers, and hints
def generate_flashcards(topic, num_flashcards=5):

    # Define the function parameters schema
    flashcard_function = {
        "name": "generate_flashcards",
        "description": "Generate flashcards with questions, answers, and hints in a structured JSON array.",
        "parameters": {
            "type": "object",
            "properties": {
                "flashcards": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "question": {"type": "string"},
                            "answer": {"type": "string"},
                            "hint": {"type": "string"}
                        },
                        "required": ["question", "answer", "hint"]
                    }
                }
            },
            "required": ["flashcards"]
        }
    }

    messages = [
        {"role": "system", "content": "You are a helpful assistant that generates flashcards."},
        {"role": "user", "content": f"Generate {num_flashcards} flashcards about {topic}."}
    ]

    # Make the API request using the new format
    response = client.chat.completions.create(
        model="gpt-4-turbo",  # Replace with the correct model; "gpt-4o-mini" is not valid
        messages=messages,  # Messages should come before the function calls
        functions=[flashcard_function],  # Provide the function definition to ensure JSON output
        function_call={"name": "generate_flashcards"},  # Invoke the specific function
        max_tokens=500,  # Adjust based on expected length of response
        temperature=0.5  # Temperature can be adjusted to control randomness
    )


    # Process the response text as JSON
    try:
        # Extract the arguments from the function call response
        flashcards_json = response.choices[0].message.function_call.arguments
        flashcards = json.loads(flashcards_json)
        return flashcards
    except json.JSONDecodeError:
        return {"error": "Failed to parse the response as JSON."}
    except Exception as e:
        return {"error": str(e)}  # Catch any other unexpected errors

# Example usage
topic = "Science"
flashcards = generate_flashcards(topic, num_flashcards=5)
print(flashcards)
